lib/foodsoft_config.rb

```diff
@@ -257,6 +259,9 @@ def get_default_config
         tasks_period_days: 7,
         tasks_upfront_days: 49,
         shared_supplier_article_sync_limit: 200,
+        host: 'localhost',
+        port: '3000',
+        protocol: 'http',
```

Comment:
I'm a bit reluctant to put these defaults here, as they're really development options.
What about skipping setting these on actionmailer whe unset, leaving it to the environment config?

Reply:
the rationale was that these are indeed the defaults, as specified in `app_config.yml_EXAMPLE` - adding them here made that more consistent (all commented out settings in the config are the defaults)

i suppose, i am more keen to put fail-safe defaults because it makes it easier to get up and running.  i do agree that in every production environment, you would want to change these.  

there is a bit of a predicament here in the current project, some config is in `environments/production.rb`, some is in `app_config.yml` and maybe some others in locale or the db.

Reply:
Yes, things changed a bit over time, especially since the introduction of db-based configuration.
The hoster needs to be able to set most things in `app_config.yml` for multi-coop setups. With docker as the preferred deployment method, environment settings are unlikely to need updating (but in some cases the hoster may still want to adapt these).

For single-coop setups, the environment setup may make more sense from a Rails point of view (which defines the localhost:3000 default). That the single-coop setup can also configure the mailer settings, is a bit of a side-effect from the multi-coop support, I would say.

So I'd rather not have `foodsoft_config.rb` meddle with the mailer settings, unless overridden specifically in `app_config.yml` (which has these settings commented out by default).

Reply:
sure, that makes sense.  i think it is important to have 'one source of truth' for all configurations - it can be a nightmare to figure out what is wrong when there are several schemes being used inconsistently (config, env, and db).   sure, you can mix these, but to make it clear, perhaps all values should be fetched from `FoodsoftConfig` instead of env vars.  my understanding from what you have suggested is that single-coop installs should use env vars, but multi-coop installs should use app_config - if i misunderstood, my apologies.

i've worked with quite a few multi-tenant apps before and it does add some complexity.  as i suggested here ( https://github.com/foodcoops/foodsoft/issues/634#issuecomment-475986738 ) it would probably be better to try and remove the 'single tenant' deployment so that we're all using the same code paths.  single tenant can just be a multi-coop but with only one co-op defined. 

aside from the benefit in less branches in the code, this also makes it easier for any co-ops that deploy it to share their installation with other local co-ops too.

although, i don't have any experience with multi-coop being turned on, so perhaps i am asking for trouble.  i believe both @wvengen and @paroga are using mutli-coop turned on, right?

btw - i have also written some ansible deployment scripts as an alternative to docker.  i was getting very annoyed with docker (failing to download) and my own hosting is now on a VPS (linode) (after being on heroku for the past few years).  these aren't very well tested yet, but i can commit at some point.

Reply:
> my understanding from what you have suggested is that single-coop installs should use env vars, but multi-coop installs should use app_config

Deployment-related settings are configured in the Rails environment, or in `app_config.yml`. To make is easy to deploy Foodsoft without having to change the source code, some of these things can be set from the shell environment. This is also the Rails default (e.g. `DATABASE_URL`), and makes it possible to deploy stock Foodsoft on Heroku and Docker.

Settings that a foodcoop may change on the fly that don't depend on the hosting environment would generally be configured in the database, from the user-interface. A default may be set in `app_config.yml`, e.g. as default settings for a new coop. A foodcoop sysadmin may also want to maintain settings in the configuration file for reference (but the database always wins, unless the settings are protected).

A hoster can make settings restricted, meaning that they cannot be overridden from within Foodsoft. By default only deployment-related settings are restricted.

To make it more concrete, deployment-related settings are things like whether pages should require SSL or not, or how mail is sent, or where the database and redis can be found. Foodcoop settings are things like the margin, whether to use messages or not, etc.

So the basic philosophy could be summarized as:
- Deployment settings are done the Rails way where possible.
- Some common deployment settings can be set using environment variables.
- Deployment settings that may differ for multicoop instances are necessarily in the app config.
- Foodcoop configuration is done in the database, but hosters can provide defaults and/or lock certain settings from the app config.

> single tenant can just be a multi-coop but with only one co-op defined.

Yes, I'm wholeheartedly for this.

> ansible deployment scripts

I'm sorry you had issues with using Docker, good you found a different solution. I'm not sure where yet, but I think it would be useful to share.